File: Problemset/1019-Next Greater Node In Linked List/Stack.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        curr = head
        vals = []

        while curr is not None:
            vals.append(curr.val)
            curr = curr.next

        stack = []
        ret = [0] * len(vals)

        for i in range(len(vals) - 1, -1, -1):
            while stack != [] and vals[i] >= stack[-1]:
                stack.pop()
            if stack != []:
                ret[i] = stack[-1]
            stack.append(vals[i])

        return ret

File: Problemset/1019-Next Greater Node In Linked List/Stack.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {Integer[]}
def next_larger_nodes(head)
  curr = head
  vals = []

  until curr.nil?
    vals.push(curr.val)
    curr = curr.next
  end

  stack = []
  ret = [0] * vals.size

  (vals.size - 1..0).step(-1).each do |i|
    stack.pop until stack.empty? || vals[i] < stack[-1]
    ret[i] = stack[-1] unless stack.empty?
    stack.push(vals[i])
  end

  ret
end
